### PR TITLE
Propagate memory space from MLIR to the compiler on GPUs

### PR DESCRIPTION
**Pull request by jaro-sevcik**
Let us propagate memory space annotations on parameters and result from MLIR module to Hlo using existing utils functions (which are currently only used by TPUs). While we are there, we also hook up the layout mode annotations.

### CODE REVIEW

File: xla/pjrt/pjrt_stream_executor_client.cc
Comment (yashk2810) [inline xla/pjrt/pjrt_stream_executor_client.cc:3420]: why is it called `arg_shapes_and_layouts`? It includes the results too right? Call it something like: `arg_and_result_layouts`?

Comment (jaro-sevcik) [reply xla/pjrt/pjrt_stream_executor_client.cc]: Results are buried somewhere in [build options](https://github.com/openxla/xla/blob/f052a450327bc4fb73dfd3685b6e4d176859d157/xla/pjrt/utils.cc#L705). Updated the name to `arg_layouts_and_pointers`.

**Comment by jaro-sevcik** [comment]
@yashk2810 Could you take another look, please?

File: xla/pjrt/pjrt_stream_executor_client.cc
Comment (yashk2810) [inline xla/pjrt/pjrt_stream_executor_client.cc:3488]: What's `build_options.result_layout`? The out_layouts we care about are the layouts and memory_spaces you calculated on line 3470 and 3466.

Comment (jaro-sevcik) [reply xla/pjrt/pjrt_stream_executor_client.cc:3488]: For some reason, the LayoutModesToXla function stores the result into build_options, see here: https://github.com/openxla/xla/blob/5e5a794c256b91695f90f7502185aa56cf80b66c/xla/pjrt/utils.cc#L705

The layouts calculated on line 3470 and 3466 need to be converted to XLA, that's what LayoutModesToXla function is doing.

Comment (jaro-sevcik) [reply xla/pjrt/pjrt_stream_executor_client.cc:3488]: TBH, the LayoutModesToXla function has some weird contract, maybe we should be using  LayoutModesToXlaShapes?

File: xla/pjrt/pjrt_stream_executor_client.cc
Comment (jyingl3) [inline xla/pjrt/pjrt_stream_executor_client.cc:3474]: If LayoutModesToXla always updates options.build_options.result_layout() in place, you do not need to pass result_layout to Compile, and read it from options instead. It does change the assumption of the newly added Compile method but it may be more consistent if you are using LayoutModesToXla. What do you think?

Comment (jaro-sevcik) [reply xla/pjrt/pjrt_stream_executor_client.cc:3474]: The original version of Compile seems to go out of its way to keep the options intact (it takes a copy of the options early on and that unchanged copy is stored into the executable), so I was concerned that changing the options might have some bad consequences. If you think it's ok to mutate the options, I am happy to take that path.

**Comment by jaro-sevcik** [comment]
@jyingl3 Could you take another look?

I have removed the result_layout parameter (passing the result_layout in options) and made the new Compile method variant protected.

File: xla/pjrt/pjrt_stream_executor_client.cc
Comment (jaro-sevcik) [reply xla/pjrt/pjrt_stream_executor_client.cc]: Removed now.

File: xla/pjrt/pjrt_stream_executor_client.cc
Comment (jaro-sevcik) [reply xla/pjrt/pjrt_stream_executor_client.cc]: [Resolving, let's stick to LayoutModesToXla.]

File: xla/pjrt/pjrt_stream_executor_client.cc
Comment (jyingl3) [inline xla/pjrt/pjrt_stream_executor_client.cc:3474]: Add a comment this method will update the result_layout in options.executable_build_options?

I will approve right after the comment is added (cannot approve right now as it will trigger pulling of the change)

**Comment by jyingl3** [comment]
> @jyingl3 Could you take another look?
> 
> I have removed the result_layout parameter (passing the result_layout in options) and made the new Compile method variant protected.

Thank you! Just one more minor suggestion about adding some comments. Will approve as soon as the comment is added.

**Comment by jaro-sevcik** [comment]
Thank you for the review, I've added the comment.

**Review by jyingl3 — APPROVED** [review]

**Review by reedwm — CHANGES_REQUESTED** [review]
I can't yet merge since there are two test failures.

In `se_gpu_pjrt_client_test.cc`, `TF_ASSIGN_OR_RETURN` is called, but tests cannot return a Status. Use `TF_ASSERT_OK_AND_ASSIGN` instead.

The test `xla_client_test.py` is failing with

```
Traceback (most recent call last):
  File "/build/work/8a6b11f4211fb6bfb014b3e6f705065fc499/google3/runfiles/google3/third_party/tensorflow/compiler/xla/python/xla_client_test.py", line 799, in testSetArgumentLayoutsLegacy
    self.assertEqual(
AssertionError: Tuples differ: (2, 1, 0) != (1, 2, 0)

First differing element 0:
2
1

- (2, 1, 0)
+ (1, 2, 0)
``` 

@jaro-sevcik, can you fix these issues?

**Comment by jaro-sevcik** [comment]
I have fixed the TF_ASSIGN_OR_RETURN and attempted a fix of the legacy layouts.

Unfortunately, I have not managed to run the xla_client_test.py test, but I have reproduced the problem with a similar C++ test (and added layout tests to this PR).

Could you please try again?

**Review by akuegel — APPROVED** [review]

**Pull request closed without merging**
